lib/recursive-methods.rb

```diff
@@ -1,49 +1,94 @@
+# frozen_string_literal: true
+
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+
+  return 1 if n <= 1
+
+
+  n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
```

Comment:
👍 Given how you have the `# frozen_string_literal: true` which makes the string fixed and means the slice isn't recreating the string.